promql/engine.go

```diff
@@ -697,117 +655,345 @@ func (ev *evaluator) eval(expr Expr) Value {
+			} else {
+				it.Reset(s.Iterator())
+			}
+			ss := Series{
+				Metric: e.series[i].Labels(),
+				Points: getPointSlice(numSteps),
+			}
 
-	case *UnaryExpr:
-		se := ev.evalOneOf(e.Expr, ValueTypeScalar, ValueTypeVector)
-		// Only + and - are possible operators.
-		if e.Op == itemSUB {
-			switch v := se.(type) {
-			case Scalar:
-				v.V = -v.V
-			case Vector:
-				for i, sv := range v {
-					v[i].V = -sv.V
+			for ts := ev.Timestamp; ts <= ev.EndTimestamp; ts += ev.Interval {
+				_, v, ok := ev.vectorSelectorSingle(it, e, ts)
+				if ok {
+					ss.Points = append(ss.Points, Point{V: v, T: ts})
 				}
 			}
+
+			if len(ss.Points) > 0 {
+				mat = append(mat, ss)
+			}
 		}
-		return se
+		return mat
 
-	case *VectorSelector:
-		return ev.vectorSelector(e)
+	case *MatrixSelector:
+		if ev.Timestamp != ev.EndTimestamp && ev.Interval != 0 {
+			panic(fmt.Errorf("cannot do range evaluatoin of matrix selector"))
+		}
+		return ev.matrixSelector(e)
 	}
+
 	panic(fmt.Errorf("unhandled expression of type: %T", expr))
 }
 
 // vectorSelector evaluates a *VectorSelector expression.
-func (ev *evaluator) vectorSelector(node *VectorSelector) Vector {
+func (ev *evaluator) vectorSelector(node *VectorSelector, ts int64) Vector {
 	var (
-		vec     = make(Vector, 0, len(node.series))
-		refTime = ev.Timestamp - durationMilliseconds(node.Offset)
+		vec = make(Vector, 0, len(node.series))
 	)
 
-	for i, it := range node.iterators {
-		var t int64
-		var v float64
-
-		ok := it.Seek(refTime)
-		if !ok {
-			if it.Err() != nil {
-				ev.error(it.Err())
-			}
+	var it *storage.BufferedSeriesIterator
+	for i, s := range node.series {
+		if it == nil {
+			it = storage.NewBuffer(s.Iterator(), durationMilliseconds(LookbackDelta))
+		} else {
+			it.Reset(s.Iterator())
 		}
 
+		t, v, ok := ev.vectorSelectorSingle(it, node, ts)
 		if ok {
-			t, v = it.Values()
+			vec = append(vec, Sample{
+				Metric: node.series[i].Labels(),
+				Point:  Point{V: v, T: t},
+			})
 		}
 
-		peek := 1
-		if !ok || t > refTime {
-			t, v, ok = it.PeekBack(peek)
-			peek++
-			if !ok || t < refTime-durationMilliseconds(LookbackDelta) {
-				continue
-			}
-		}
-		if value.IsStaleNaN(v) {
-			continue
+	}
+	return vec
+}
+
+func (ev *evaluator) vectorSelectorSingle(it *storage.BufferedSeriesIterator, node *VectorSelector, ts int64) (int64, float64, bool) {
+	refTime := ts - durationMilliseconds(node.Offset)
+	var t int64
+	var v float64
+
+	ok := it.Seek(refTime)
+	if !ok {
+		if it.Err() != nil {
+			ev.error(it.Err())
 		}
+	}
 
-		vec = append(vec, Sample{
-			Metric: node.series[i].Labels(),
-			Point:  Point{V: v, T: t},
-		})
+	if ok {
+		t, v = it.Values()
 	}
-	return vec
+
+	peek := 1
+	if !ok || t > refTime {
+		t, v, ok = it.PeekBack(peek)
+		peek++
+		if !ok || t < refTime-durationMilliseconds(LookbackDelta) {
+			return 0, 0, false
+		}
+	}
+	if value.IsStaleNaN(v) {
+		return 0, 0, false
+	}
+	return t, v, true
 }
 
 var pointPool = sync.Pool{}
```

Comment:
We have a size-based pool package now. Probably makes sense to use that.

https://github.com/prometheus/prometheus/tree/master/pkg/pool

Can be another PR of course.

Reply:
Given how the pools are used I'm not sure that'd actually be a win, it'd take a fair bit of operational experience to determine that.